auth Lua health checks: more responsiveness

miodvallat opened the pull request:
### Short description
The current state of the health check thread is to check for statuses checks to launch, every `lua-health-checks-interval` seconds.

This design has two issues:
- if the thread runs before its work list is filled, then the checks won't be started until `lua-health-checks-interval` seconds, which will be noticeable if this value is large (as noticed in #16369).
- processing Lua records causing new health checks to be started won't cause them to be noticed and work on until the thread wakes up from its `lua-health-checks-interval` seconds sleep.

This PR attempts to address this in two commits:
- the checker thread was always being started (if not running already) before adding checks to its work list, creating a race where the thread could run, notice its list of checks is empty, and go to sleep, after which its check job is immediately added.
- because further check requests can occur at any time, let it wake up every second to check for more work to do.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] read and accepted the [Developer Certificate of Origin](https://github.com/PowerDNS/pdns/blob/master/DCO) document, including the [AI Policy](https://github.com/PowerDNS/pdns/blob/master/AI_POLICY.md), and added a ["Signed-off-by"](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md#developer-certificate-of-origin) to my commits
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

miodvallat committed:
Create the lua health check thread after giving it work to do, not before.

Signed-off-by: Miod Vallat <miod.vallat@powerdns.com>

miodvallat committed:
Make sure the health check thread checks for new work every second.

Fixes: #16369

Signed-off-by: Miod Vallat <miod.vallat@powerdns.com>

Habbie commented on pdns/lua-record.cc:
```diff
@@ -259,7 +259,10 @@ class IsUpOracle
       setThreadName("pdns/luaupcheck");
 
-      std::this_thread::sleep_until(checkStart + std::chrono::seconds(interval));
+      // Only sleep for 1 second here, even if the health check interval is
```
would it make sense to instead have a longer, but interruptible, sleep, where something that gives us work wakes us up? It would save on thread wakeups, but it might be more expensive in other ways.

miodvallat replied:
I had considered this, but this will require a conditional variable on which to sleep, and its associated mutex, and I'm not sure this will keep the code readable. Let me tinker a bit with that idea...

Habbie reviewed (dismissed):
approved, but please think about the comment i posted

miodvallat committed:
Use a condition variable to signal the checker thread.

Signed-off-by: Miod Vallat <miod.vallat@powerdns.com>

miodvallat committed:
Use different IP addresses for the various unreachability tests.

Otherwise, the periodic checks from the tests which have run previously
can still trigger and change the status of a given IP at a different time
than expected by the test code. This is especially true in CI environments.

Signed-off-by: Miod Vallat <miod.vallat@powerdns.com>

miodvallat commented:
If/when backporting, #17514 will likely need backporting too.
